torch_xla/runtime.py

```diff
@@ -0,0 +1,236 @@
+import functools
+import logging
+import os
+from typing import Dict, List, Optional, TypeVar
+
+import torch
+import torch_xla
+import torch_xla.core.xla_env_vars as xenv
+import torch_xla.core.xla_model as xm
+import torch_xla.distributed.xla_backend
+import torch_xla.utils.utils as xu
+
+R = TypeVar('R')
+FN = TypeVar('FN')
+
+
+def set_device_type(pjrt_device: str) -> None:
+  """Sets the current PjRt device type.
+
+  Must be run before using any XLA devices.
+
+  Args:
+    pjrt_device: 'TPU' or 'CPU'
+  """
+  os.environ[xenv.PJRT_DEVICE] = pjrt_device
+
+
+def _maybe_select_default_device():
+  # Skip if runtime is already configured
+  if xu.getenv_as(
+      xenv.PJRT_SELECT_DEFAULT_DEVICE, str, '1'
+  ) == '0' or xenv.PJRT_DEVICE in os.environ or xenv.GPU_NUM_DEVICES in os.environ or any(
+      env.startswith('XRT_') for env in os.environ):
+    return
+
+  logging.warning(
+      'XRT configuration not detected. Defaulting to preview PJRT '
+      'runtime. To silence this warning and continue using PJRT, '
+      'explicitly set PJRT_DEVICE to a supported device or configure XRT. To '
+      'disable default device selection, set PJRT_SELECT_DEFAULT_DEVICE=0')
+  # TODO: Update this link in the release branch
+  logging.warning('For more information about the status of PJRT, see '
+                  'https://github.com/pytorch/xla/blob/master/docs/xr.md')
+  # Check for libtpu _and_ the TPU device
+  if torch_xla._found_libtpu and os.path.exists('/dev/accel0'):
+    logging.warning('libtpu.so and TPU device found. Setting PJRT_DEVICE=TPU.')
+    os.environ[xenv.PJRT_DEVICE] = 'TPU'
+  else:
+    logging.warning('Defaulting to PJRT_DEVICE=CPU')
+    os.environ[xenv.PJRT_DEVICE] = 'CPU'
+  # TODO(wcromar): Detect GPU device too
+
+
+def device_type() -> Optional[str]:
```

Comment:
I think we have similar functions in `torch_xla.core.xla_model`, do we want to do some clean up?

Reply:
I want to take this chance to do some clean up. I am always confuse what function to call for `local ordinal`, `gloabal ordinal`, `worla_size` etc and what do they really mean in a pod context. If we can restructure those api a bit and maybe have those apis in this runtime instead that would be nice...(random idea, might need more thinking)

Reply:
`xla_model` becomes a kitch sink and we put random things in it, if we can move all runtime related bits in this file it is actually nicer..

Reply:
Discussed this offline. We'll start to move APIs that interact directly with the runtime to a new module, and leave any modeling-related APIs in xla_model. I moved the PJRT version of rendezvous back to xla_model, and the old `rendezvous` is will be an alias of that implementation when PJRT is enabled.